core/store/ocr_key_store.go

```diff
@@ -0,0 +1,145 @@
+package store
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/pkg/errors"
+	"go.uber.org/multierr"
+
+	"github.com/smartcontractkit/chainlink/core/store/models/ocrkey"
+)
+
+// OCRKeyStore holds the OCR keys in memory
+type OCRKeyStore struct {
+	lock  sync.RWMutex
+	keys  InMemoryOCRKeyStore
+	store *Store
+}
+
+// InMemoryOCRKeyStore holds all the unlocked OCRPrivateKey, mapped by the IDs of their
+// encrypted DB records
+type InMemoryOCRKeyStore = map[int32]*ocrkey.OCRPrivateKey
+
+// NewOCRKeyStore returns an empty OCRKeyStore
+func NewOCRKeyStore(store *Store) *OCRKeyStore {
+	return &OCRKeyStore{
+		lock:  sync.RWMutex{},
+		keys:  make(InMemoryOCRKeyStore),
+		store: store,
+	}
+}
+
+// CreateKey creates a new private key, persists it to the DB, and adds it
+// to the in-memory keystore
+func (ks *OCRKeyStore) CreateKey(auth string) (*ocrkey.OCRPrivateKey, error) {
+	return ks.createKey(auth, ocrkey.DefaultScryptParams)
+}
+
+// CreateWeakKeyXXXTestingOnly creates a new private key with weak encryption parameters and should
+// only be used for testing
+func (ks *OCRKeyStore) CreateWeakKeyXXXTestingOnly(auth string) (*ocrkey.OCRPrivateKey, error) {
+	return ks.createKey(auth, ocrkey.FastScryptParams)
+}
+
+func (ks *OCRKeyStore) createKey(auth string, scryptParams ocrkey.ScryptParams) (*ocrkey.OCRPrivateKey, error) {
+	ks.lock.Lock()
+	defer ks.lock.Unlock()
+	key, err := ocrkey.NewOCRPrivateKey()
+	if err != nil {
+		return nil, err
+	}
+	if err = ks.saveToDB(key, auth, scryptParams); err != nil {
+		return nil, err
+	}
+	if err = ks.addToKeystore(key); err != nil {
+		return nil, err
+	}
+	return key, nil
+}
+
+// Unlock tries to unlock each ocr key in the db, using the given pass phrase,
+// and returns any keys it manages to unlock. It returns an error if none were unlocked.
+func (ks *OCRKeyStore) Unlock(phrase string) (keysUnlocked []int32, merr error) {
+	ks.lock.Lock()
```

Comment:
This mutex wraps a database call, this could block for a long time.

Can we scope it to be smaller?